Fix #696 complicated delete problem: continues #700

Shougo opened the pull request:
It fixes #696.
It is based on #700.

I have fixed `test_delete.vim` tests.

@k-takata : Can you test it in Windows?
It seems better #700.  But it may not work in Windows.
Any fix is appreciable.

Shougo committed:
Fix #696 complicated delete problem: continues #700

k-takata commented:
This is actually a difficult problem.
After I wrote the comments in #700, I found that replacing `[` with `[[]` doesn't always work.

If there is a file named `foobar[.txt`, `:w! foobar[[].txt` writes the current buffer to the file `foobar[.txt`.
However if there is no file named `foobar[.txt`, `:w! foobar[[].txt` writes to the file `foobar[[].txt` instead of `foobar[.txt`. The behavior depends on the existing files.
This means that there is no general way to escape a filename on Windows.

I think there are (at least) two ways to fix the problem.
1. Add a new flag to `gen_expand_wildcards()` (or create a new function) to list all files and directories under the specified directory.  
   Currently, we pass "pathname/*" to `gen_expand_wildcards()` but we don't need to expand the part of "pathname/".
2. Create a general specification to escape a filename on Windows, then use it for this.  
   I think this is difficult, and this could be cause some backward incompatibilities.

h-east commented on src/fileio.c:
```diff
@@ -7261,7 +7261,16 @@ delete_recursive(char_u *name)
+	/* Escape some regexp chars in file names as these probably will be
+	 * interpreted as ill-formed regular expressions in
+	 * gen_expand_wildcards. */
+	esc_name = do_string_sub(name, (char_u *)"[",
```
I don't know this patch is correct.
But, You should do the NULL check to the esc_name and esc_name2.
If NULL, call the vim_free() properly and return -1.

Shougo commented:
> I don't know this patch is correct.
> But, You should do the NULL check to the esc_name and esc_name2.
> If NULL, call the vim_free() properly and return -1.

Thank you.  I have added the NULL check.

And I have fixed the tests error.

> Add a new flag to gen_expand_wildcards() (or create a new function) to list all files and directories under the specified directory.

I think it is better.  But it seems difficult...

k-takata commented:
> I think it is better. But it seems difficult...

Yeah, agree. But I don't think we can fix this without it.
https://ci.appveyor.com/project/chrisbra/vim/build/2031/job/lcfupaxg9gd50kfp#L2019

```
Found errors in Test_complicated_name_recursive_delete():
function RunTheTest[9]..Test_complicated_name_recursive_delete line 14: Expected False but got 1
TEST FAILURE 
```

Shougo commented:
We need listdir() implementation.
I may implement it for Unix.
But it is difficult to implement for Windows.

k-takata commented:
Fixed by [8.1.1378](https://github.com/vim/vim/commit/701ff0a3e53d253d7300c385e582659bbff7860d).
